### ISSUE TITLE
How to change loss function, without installing ultralytics packages

### ISSUE THREAD

**Issue by TikPapyan** [ultralytics/ultralytics#6065]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to make some changes to tuning, specifically changing the loss function, but it doesn't get inside it when running. I didn't install ultralytics, hence the yolov5 packages, coming with it, because the error I am now getting disappears, but the changes made on the loss function don't have an impact. Running the code without installed packages returns this error:

UnboundLocalError                         Traceback (most recent call last)
Cell In[5], line 1
----> 1 model.tune(data="/datasets/caltechpedestriandataset.yaml", 
      2            epochs=50, iterations=300, optimizer='AdamW', plots=False, save=False, val=False)

File ~/.local/lib/python3.10/site-packages/ultralytics/engine/model.py:366, in Model.tune(self, use_ray, iterations, *args, **kwargs)
    364 custom = {}  # method defaults
    365 args = {**self.overrides, **custom, **kwargs, 'mode': 'train'}  # highest priority args on the right
--> 366 return Tuner(args=args, _callbacks=self.callbacks)(model=self, iterations=iterations)

File ~/.local/lib/python3.10/site-packages/ultralytics/engine/tuner.py:207, in Tuner.__call__(self, model, iterations, cleanup)
    205         shutil.copy2(ckpt, self.tune_dir / 'weights')
    206 elif cleanup:
--> 207     shutil.rmtree(ckpt_file.parent)  # remove iteration weights/ dir to reduce storage space
    209 # Plot tune results
    210 plot_tune_results(self.tune_csv)

UnboundLocalError: local variable 'ckpt_file' referenced before assignment

Installing the packages (yolov5) and adding a simple exit code to the loss function:

class BboxLoss(nn.Module):
    """Criterion class for computing training losses during training."""

    def __init__(self, reg_max, use_dfl=False):
        """Initialize the BboxLoss module with regularization maximum and DFL settings."""
        super().__init__()
        self.reg_max = reg_max
        self.use_dfl = use_dfl

    def forward(self, pred_dist, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask):
        """IoU loss."""
        print("Inside loss function")
        exit(0)

And it doesn't take this into account - it runs as normal and doesn't print the message or exit the function.

The question is how can I modify the loss function? Do I have to do something with the setup.py?

**Comment by glenn-jocher** [comment]
@TikPapyan hi there,

Since you're attempting to modify the loss function without using Ultralytics packages, you have to confirm that the changes you are making are being utilized. In Python, when you make changes to a module that's already been imported or used, the changes aren't necessarily reflected in the already running instance. You would have to reload the module for the changes to take effect. The built-in `reload()` function in Python could be of help in this case.

However, it is important to note that changing the loss function can be complex and is not necessarily a straightforward process. The loss function of YOLOv8 was designed with specific behaviors and complexities in mind, and altering it could lead to unexpected results or errors.

To summarize, to change the loss function:
1. Make your desired changes to the loss function file.
2. Ensure these modifications are imported and used by the main scripts.
3. You may have to use the `reload()` function in Python to make sure the changes are reflected.

Bear in mind that the Ultralytics team cannot promise support for individual customizations to the code such as this, but I hope that this direction may help you to achieve your goals.
